### PR TITLE
add build doc

### PR DESCRIPTION
**Pull request by Yurlungur**
## PR Summary

Document how to build `singularity-eos` in all its complicated glory.

## PR Checklist

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- [x] Document any new features, update documentation for changes made.

### CODE REVIEW

**Comment by jhp-lanl** [comment]
Thanks for working on this!

File: doc/sphinx/src/building.rst [hunk 1]
```diff
@@ -3,3 +3,216 @@
...
+
+.. _Spack: https://spack.io/
+
+.. code-block:: bash
+  git clone --recursive git@github.com:lanl/singularity-eos.git
+  spack repo add singularity-eos/spack-repo
+  spack install singularity-eos
+
+to install ``singularity-eos`` into your spack instance. The spackage
+supports a number of relevant variants:
```
Comment (annapiegra) [inline doc/sphinx/src/building.rst:26]: Is the spackage ready for general use:
I tried: 
`spack install singularity-eos +kokkos +tests build_extra=sesame`
and got:
`py-h5py requires hdf5 variant +mpi, but spec asked for ~mpi`

Comment (mauneyc-LANL) [reply doc/sphinx/src/building.rst:26]: Is mpi an external dependency or being scheduled to build by spack? i.e. is there an external, non-buildable mpi in a package.yaml file? Also, is this in the context of an environment? there may be something in `$HOME/.spack/` that is conflicting here.

File: doc/sphinx/src/building.rst [hunk 2]
```diff
@@ -3,3 +3,216 @@
...
++-----------------------------+---------+-----------------+-----------------------------+
+| kokkos-kernels [off]        | -----   | on, off         | Use kokkos-kernels for      |
+|                             |         |                 | linear algebra suport,      |
+|                             |         |                 | which is needed with        |
+|                             |         |                 | mixed-cell closures on GPU  |
++-----------------------------+---------+-----------------+-----------------------------+
+| mpi [off]                   | -----   | on, off         | Build with parallel HDF5    |
+|                             |         |                 | otherwise build with serial |
++-----------------------------+---------+-----------------+-----------------------------+
+| opemp [off]                 | +kokkos | on, off         | Build Kokkos openmp backend |
```
Comment (dholladay00) [inline doc/sphinx/src/building.rst:61]: `openmp`

File: doc/sphinx/src/building.rst [hunk 3]
```diff
@@ -3,3 +3,216 @@
...
+.. _kokkos-kernels: https://github.com/kokkos/kokkos-kernels/
+
+If you use spack, but would like to build ``singularity-eos`` from
+source, you can install dependencies via, e.g.,
+
+.. code-block:: bash
+  git clone --recursive git@github.com:lanl/singularity-eos.git
+  spack repo add singularity-eos/spack-repo
+  spack install --only dependencies singularity-eos+cuda cuda_arch=70
+
```
Comment (dholladay00) [inline doc/sphinx/src/building.rst:217]: If I have a checkout of singularity-eos and I want to build in a non-spack build directory, is there still a way to get the cmake options to match the spack options @mauneyc-LANL ? When there was a deps package, there was a cmake configuration file generated to aid that process, does such a thing still exist, or can we specify via spack where we want the whole singularity-eos spackage to be built?

File: doc/sphinx/src/building.rst (hunk 1 above)
Comment (annapiegra) [reply doc/sphinx/src/building.rst:26]: Removing ~/.spack and adding +mpi concretized (will not concretize w/o +mpi) and currently building.

Comment (mauneyc-LANL) [reply doc/sphinx/src/building.rst:26]: I will update the spackage to conflict if `+sesame~mpi`, which seems to be causing the issue (or whatever else variant requires `py-hdf5`

Comment (Yurlungur) [reply doc/sphinx/src/building.rst:26]: don't know why hdf5 requires MPI. It shouldn't. But... 🤷 python is always finicky.

File: doc/sphinx/src/building.rst (hunk 3 above)
Comment (Yurlungur) [reply doc/sphinx/src/building.rst:217]: This is a good question that I would also like to know.

**Comment by Yurlungur** [comment]
@dholladay00 @annapiegra @mauneyc-LANL there seem to be some questions regarding the spackage. That said, I'd like to bring the documentation into the code regardless of any changes/fixes to the spack that may be needed. How do you all feel about that?

File: doc/sphinx/src/building.rst (hunk 3 above)
Comment (ktsai7) [reply doc/sphinx/src/building.rst:218]: spack has a `install_tree` config that we could modify on a user level, but I don't know if that's the best approach you guys want here

File: doc/sphinx/src/building.rst (hunk 1 above)
Comment (ktsai7) [reply doc/sphinx/src/building.rst:26]: @Yurlungur I think `py-hdf5` is needed otherwise something doesn't work when we were first adding the CI?

Comment (Yurlungur) [reply doc/sphinx/src/building.rst:27]: Right `py-hdf5` is needed for the `stellarcollapse` tests. What confuses me is that it needs `mpi`. There's no reason to build parallel hdf5 for python.

File: doc/sphinx/src/building.rst (hunk 3 above)
Comment (mauneyc-LANL) [reply doc/sphinx/src/building.rst:220]: This feature is not included in the `package.py`, tho I can't recall why I cut it. 

As for controlling where it's build, that would require a global configuration change. Something like

```bash
$ spack location --install-dir gcc
~/spack/opt/spack/linux-centos7-x86_64/gcc-4.9.3-iy4rw...
```

would be helpful in the short term. Long-term, probably something like what `-deps` was doing seemed useful for devs, so we can return to it

**Comment by mauneyc-LANL** [comment]
> @dholladay00 @annapiegra @mauneyc-LANL there seem to be some questions regarding the spackage. That said, I'd like to bring the documentation into the code regardless of any changes/fixes to the spack that may be needed. How do you all feel about that?

Dumb question: should we be putting "sphinx"-style comments in the spackage?

**Comment by Yurlungur** [comment]
> > @dholladay00 @annapiegra @mauneyc-LANL there seem to be some questions regarding the spackage. That said, I'd like to bring the documentation into the code regardless of any changes/fixes to the spack that may be needed. How do you all feel about that?
> 
> Dumb question: should we be putting "sphinx"-style comments in the spackage?

I'm not sure that's needed... we're basically using sphinx rst to hand-write the docs, but we're not doing "doxygen" style automated documentation. I'd be open to that path though if people think it's worth it.

File: doc/sphinx/src/building.rst (hunk 3 above)
Comment (Yurlungur) [reply doc/sphinx/src/building.rst:220]: could we just have both, @mauneyc-LANL ? A deps package and a full package? And the devs can just install the deps package?

Comment (mauneyc-LANL) [reply doc/sphinx/src/building.rst:221]: We could but I don't think it's necessary. Let's discuss some options later today (if you're around). I can try to mock up something later today.

**Comment by jhp-lanl** [comment]
> > > @dholladay00 @annapiegra @mauneyc-LANL there seem to be some questions regarding the spackage. That said, I'd like to bring the documentation into the code regardless of any changes/fixes to the spack that may be needed. How do you all feel about that?
> > 
> > 
> > Dumb question: should we be putting "sphinx"-style comments in the spackage?
> 
> I'm not sure that's needed... we're basically using sphinx rst to hand-write the docs, but we're not doing "doxygen" style automated documentation. I'd be open to that path though if people think it's worth it.

I think the only advantage of using docstrings in the `package.py` file (which I think is what is being proposed?) is that the comment will exist where the code change is likely being made. Editing two files just increases the chance that the documentation will be out of date. Though as you pointed out Jonah, there's always that chance since the documentation isn't auto-generated so the advantage may not be significant.

**Comment by Yurlungur** [comment]
> > > > @dholladay00 @annapiegra @mauneyc-LANL there seem to be some questions regarding the spackage. That said, I'd like to bring the documentation into the code regardless of any changes/fixes to the spack that may be needed. How do you all feel about that?
> > > 
> > > 
> > > Dumb question: should we be putting "sphinx"-style comments in the spackage?
> > 
> > 
> > I'm not sure that's needed... we're basically using sphinx rst to hand-write the docs, but we're not doing "doxygen" style automated documentation. I'd be open to that path though if people think it's worth it.
> 
> I think the only advantage of using docstrings in the `package.py` file (which I think is what is being proposed?) is that the comment will exist where the code change is likely being made. Editing two files just increases the chance that the documentation will be out of date. Though as you pointed out Jonah, there's always that chance since the documentation isn't auto-generated so the advantage may not be significant.

Also for the purpose of building the code, commenting the code in the spackage probably isn't useful. All we actually care about is what spack variants are available in the spackage. Though maybe there's a way to automatically generate the table I hand-wrote here, which would be nice.

File: doc/sphinx/src/building.rst (hunk 3 above)
Comment (Yurlungur) [reply doc/sphinx/src/building.rst:221]: yep sounds good

Comment (ktsai7) [reply doc/sphinx/src/building.rst:221]: there is the option of having `singularity-eos-deps` spackage and inherit from it to make the `singularity-eos` spackage... at least you won't have to maintain in two places

File: doc/sphinx/src/building.rst (hunk 1 above)
Comment (Yurlungur) [reply doc/sphinx/src/building.rst:27]: Ah I think I found the problem. `py-h5py` has `mpi` on by default in its variant. Do we need to pass `~mpi` down through it?

**Comment by ktsai7** [comment]
> > > > > @dholladay00 @annapiegra @mauneyc-LANL there seem to be some questions regarding the spackage. That said, I'd like to bring the documentation into the code regardless of any changes/fixes to the spack that may be needed. How do you all feel about that?
> > > > 
> > > > 
> > > > Dumb question: should we be putting "sphinx"-style comments in the spackage?
> > > 
> > > 
> > > I'm not sure that's needed... we're basically using sphinx rst to hand-write the docs, but we're not doing "doxygen" style automated documentation. I'd be open to that path though if people think it's worth it.
> > 
> > 
> > I think the only advantage of using docstrings in the `package.py` file (which I think is what is being proposed?) is that the comment will exist where the code change is likely being made. Editing two files just increases the chance that the documentation will be out of date. Though as you pointed out Jonah, there's always that chance since the documentation isn't auto-generated so the advantage may not be significant.
> 
> Also for the purpose of building the code, commenting the code in the spackage probably isn't useful. All we actually care about is what spack variants are available in the spackage. Though maybe there's a way to automatically generate the table I hand-wrote here, which would be nice.

It's not exactly a table, but we could pull in code into the `.rst` files so when someone update the code it would get pull in automatically...

File: doc/sphinx/src/building.rst (hunk 3 above)
Comment (Yurlungur) [reply doc/sphinx/src/building.rst:221]: Right that's what I was thinking.

Comment (dholladay00) [reply doc/sphinx/src/building.rst:221]: Definitely don't want to maintain 2 packages, but I'd like to hear what @mauneyc-LANL is thinking. I know the spack people want folks to move away from `-deps` spackages and I'm assuming they have reasons and I'd like to understand that better before going against the grain.

File: doc/sphinx/src/building.rst (hunk 1 above)
Comment (annapiegra) [reply doc/sphinx/src/building.rst:27]: Ok, spack build w/ ~mpi fails with:
  ```
   131                     from /tmp/annap/spack-stage/spack-stage-singularit
            y-eos-main-brgwdvmhfkzipj2bjry2g2e5ws6em2hj/spack-src/singularity-e
            os/eos/singularity_eos.cpp:19:
  >> 132    /tmp/annap/spack-stage/spack-stage-singularity-eos-main-brgwdvmhfkz
            ipj2bjry2g2e5ws6em2hj/spack-src/singularity-eos/closure/mixed_cell_
            models.hpp:272:2: error: #error "Eigen should not be used with Kokk
            os."
     133      272 | #error "Eigen should not be used with Kokkos."
     134          |  ^~~~~
 
```

and 

```
     141    /usr/bin/ld: /projects/artab/users/annap/spack-environments/DWP9X/1
            .8.0/spack/opt/spack/linux-rhel7-power9le/gcc-9.3.0/eospac-6.4.2bet
            a-jxle75damh3twfcaicx7xdboktbyugk2/lib/libeospac6.a(eos_Interpolati
            on.o): In function `eos_BiRationalInterpolate':
     142    eos_Interpolation.c:(.text+0x520c): toc optimization is not support
            ed for 0xe4e50002 instruction.
     143    make[2]: Leaving directory `/ram/tmp/annap/spack-stage/spack-stage-
            singularity-eos-main-brgwdvmhfkzipj2bjry2g2e5ws6em2hj/spack-build-b
            rgwdvm'
     144    [ 73%] Built target sesame2spiner
  >> 145    make[2]: *** [singularity-eos/eos/CMakeFiles/eos_objs.dir/singulari
            ty_eos.cpp.o] Error 1
     146    make[2]: Leaving directory `/ram/tmp/annap/spack-stage/spack-stage-
            singularity-eos-main-brgwdvmhfkzipj2bjry2g2e5ws6em2hj/spack-build-b
            rgwdvm'
  >> 147    make[1]: *** [singularity-eos/eos/CMakeFiles/eos_objs.dir/all] Erro
            r 2
     148    make[1]: *** Waiting for unfinished jobs....

```

**Comment by annapiegra** [comment]
> 

I would add a note saying spack builds are still under development.

File: doc/sphinx/src/building.rst (hunk 1 above)
Comment (mauneyc-LANL) [reply doc/sphinx/src/building.rst:27]: @Yurlungur What was the status of this error? Did we ever decide what to do about it?

Comment (Yurlungur) [reply doc/sphinx/src/building.rst:27]: um... I don't understand why we're hitting that error. It *should* only happen if you build with Kokkos. That said that macro is wrong. I will try to push a correction to this branch.

**Comment by Yurlungur** [comment]
> > 
> 
> I would add a note saying spack builds are still under development. Besides that I think it's great to have build instructions!

I'll add that.

**Comment by dholladay00** [comment]
I'm going to approve this as it is a good starting place for building documentation. There are still issues with the spack itself and those can be reported elsewhere, the documentation itself is sufficient for now.

**Review by dholladay00 — APPROVED** [review]

**Pull request merged**